File: leetcode/0745-prefix-and-suffix-search/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
using namespace std;

class WordFilter {
    struct TrieNode {
        unordered_map<char, TrieNode*> children;
        int maxIndex = -1;
    };

    TrieNode* root;

public:
    WordFilter(vector<string>& words) {
        root = new TrieNode();
        for (int index = 0; index < (int)words.size(); index++) {
            const string& word = words[index];
            for (int k = 0; k <= (int)word.size(); k++) {
                string combo = word.substr(k) + "#" + word;
                TrieNode* node = root;
                node->maxIndex = index;
                for (char ch : combo) {
                    if (node->children.find(ch) == node->children.end()) {
                        node->children[ch] = new TrieNode();
                    }
                    node = node->children[ch];
                    node->maxIndex = index;
                }
            }
        }
    }

    int f(string prefix, string suffix) {
        string combo = suffix + "#" + prefix;
        TrieNode* node = root;
        for (char ch : combo) {
            auto it = node->children.find(ch);
            if (it == node->children.end()) return -1;
            node = it->second;
        }
        return node->maxIndex;
    }
};

File: leetcode/0745-prefix-and-suffix-search/Solution.java
import java.util.HashMap;
import java.util.Map;

class WordFilter {
    private static class TrieNode {
        Map<Character, TrieNode> children = new HashMap<>();
        int maxIndex = -1;
    }

    private final TrieNode root = new TrieNode();

    public WordFilter(String[] words) {
        for (int index = 0; index < words.length; index++) {
            String word = words[index];
            for (int k = 0; k <= word.length(); k++) {
                String combo = word.substring(k) + "#" + word;
                TrieNode node = root;
                node.maxIndex = index;
                for (char ch : combo.toCharArray()) {
                    node = node.children.computeIfAbsent(ch, c -> new TrieNode());
                    node.maxIndex = index;
                }
            }
        }
    }

    public int f(String prefix, String suffix) {
        String combo = suffix + "#" + prefix;
        TrieNode node = root;
        for (char ch : combo.toCharArray()) {
            node = node.children.get(ch);
            if (node == null) return -1;
        }
        return node.maxIndex;
    }
}
